File: scripts/update-oss-contribs.rs
#!/usr/bin/env -S cargo +nightly -Zscript
---
[package]
edition = "2024"

[dependencies]
serde = { version = "1", features = ["derive"] }
serde_json = "1"
toml_edit = { version = "0.25", features = ["serde"] }
chrono = { version = "0.4", default-features = false, features = ["clock", "std"] }
---
//! Updates the `[[oss_contribs]]` section of `src/config/work.toml` from live
//! GitHub contribution data.
//!
//! Run from the repository root:
//!
//!   ./scripts/update-oss-contribs.rs
//!
//! Requires an authenticated `gh` CLI.
//!
//! The script:
//!   1. Fetches your contributions for the past year via the GitHub GraphQL API.
//!   2. Filters out own repos, work-org repos, and repos with no PRs.
//!   3. Preserves any existing custom display names.
//!   4. Sorts by GitHub star count descending (so the most notable appear first).
//!   5. Rewrites the `[[oss_contribs]]` section in `src/config/work.toml`.

use std::collections::HashMap;
use std::path::Path;
use std::process::Command;

// MARK: Configuration

/// GitHub login whose contributions are analysed.
const GITHUB_USER: &str = "JP-Ellis";

/// Orgs/owners excluded entirely from the misc list.
///
/// Note: individual repos already in `tracked_slugs` are also excluded
/// regardless of org, so e.g. `pact-foundation/pact-python` is deduped
/// automatically without needing its org here.
const SKIP_ORGS: &[&str] = &["JP-Ellis", "hep-rs", "pactflow"];

/// Minimum number of PRs to qualify for the misc list.
const MIN_PRS: u64 = 1;

/// Path to the config file, relative to the repository root.
const CONFIG_PATH: &str = "src/config/work.toml";

// MARK: Types

#[derive(Debug, serde::Deserialize)]
struct WorkConfig {
    tracked_slugs: Vec<String>,
    #[serde(default)]
    oss_contribs: Vec<OssContrib>,
}

#[derive(Debug, serde::Deserialize)]
struct OssContrib {
    slug: String,
    name: String,
}

#[derive(Debug, Default)]
struct RepoActivity {
    commits: u64,
    prs: u64,
    is_private: bool,
}

// MARK: Main
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let config_path = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join(CONFIG_PATH)
        .canonicalize()?;

    if !config_path.exists() {
        return Err(format!("Config file not found at {CONFIG_PATH}").into());
    }

    let config_str = std::fs::read_to_string(&config_path)
        .map_err(|e| format!("Could not read {CONFIG_PATH}: {e}"))?;
    let config: WorkConfig = toml_edit::de::from_str(&config_str)
        .map_err(|e| format!("Could not parse {CONFIG_PATH}: {e}"))?;

    // Build a name-override map from the current entries.
    let existing_names: HashMap<&str, &str> = config
        .oss_contribs
        .iter()
        .map(|c| (c.slug.as_str(), c.name.as_str()))
        .collect();

    let from = year_ago_rfc3339();
    let to = now_rfc3339();
    eprintln!("Fetching contributions {from} → {to} for @{GITHUB_USER}...");

    let activity = fetch_contributions(&from, &to)?;

    // Build a set of slugs that are already in tracked_slugs (main PROJECTS).
    let tracked: std::collections::HashSet<&str> =
        config.tracked_slugs.iter().map(|s| s.as_str()).collect();

    // Filter.
    let mut filtered: Vec<String> = activity
        .iter()
        .filter(|(slug, data)| {
            let org = slug.split('/').next().unwrap_or("");
            !SKIP_ORGS.contains(&org)
                && !tracked.contains(slug.as_str())
                && !data.is_private
                && data.prs >= MIN_PRS
        })
        .map(|(slug, _)| slug.clone())
        .collect();
    filtered.sort(); // stable base order before star sort

    // Fetch star counts and sort by popularity descending.
    eprintln!("Fetching star counts for {} repos...", filtered.len());
    let stars = fetch_stars(&filtered).unwrap_or_default();
    filtered.sort_by(|a, b| {
        stars
            .get(b)
            .copied()
            .unwrap_or(0)
            .cmp(&stars.get(a).copied().unwrap_or(0))
            .then(a.cmp(b))
    });

    // Build (slug, display_name, stars) triples, preserving custom names.
    let entries: Vec<(String, String, u64)> = filtered
        .iter()
        .map(|slug| {
            let default_name = slug.split('/').nth(1).unwrap_or(slug.as_str());
            let name = existing_names
                .get(slug.as_str())
                .copied()
                .unwrap_or(default_name)
                .to_string();
            let star_count = stars.get(slug).copied().unwrap_or(0);
            (slug.clone(), name, star_count)
        })
        .collect();

    let patched = patch_oss_contribs(&config_str, &entries)?;
    std::fs::write(&config_path, &patched)
        .map_err(|e| format!("Could not write {CONFIG_PATH}: {e}"))?;

    println!(
        "Updated {CONFIG_PATH} — {} OSS contrib entries:",
        entries.len()
    );
    for slug in &filtered {
        let is_new = !existing_names.contains_key(slug.as_str());
        let star_count = stars.get(slug).copied().unwrap_or(0);
        let star_str = if star_count >= 1_000 {
            format!("{:.1}k★", star_count as f64 / 1_000.0)
        } else {
            format!("{star_count}★")
        };
        let tag = if is_new { "  [new]" } else { "" };
        println!("  {slug} ({star_str}){tag}");
    }

    Ok(())
}

// MARK: GitHub API
fn fetch_contributions(
    from: &str,
    to: &str,
) -> Result<HashMap<String, RepoActivity>, Box<dyn std::error::Error>> {
    let query = format!(
        r#"{{
  user(login: "{GITHUB_USER}") {{
    contributionsCollection(from: "{from}", to: "{to}") {{
      commitContributionsByRepository(maxRepositories: 100) {{
        repository {{ nameWithOwner isPrivate }}
        contributions {{ totalCount }}
      }}
      pullRequestContributionsByRepository(maxRepositories: 100) {{
        repository {{ nameWithOwner isPrivate }}
        contributions {{ totalCount }}
      }}
    }}
  }}
}}"#
    );

    let output = Command::new("gh")
        .args(["api", "graphql", "-f", &format!("query={query}")])
        .output()
        .map_err(|e| {
            format!("Failed to run `gh`: {e}. Is the gh CLI installed and authenticated?")
        })?;

    if !output.status.success() {
        return Err(format!(
            "`gh api graphql` failed:\n{}",
            String::from_utf8_lossy(&output.stderr)
        )
        .into());
    }

    let json: serde_json::Value = serde_json::from_slice(&output.stdout)?;
    let collection = &json["data"]["user"]["contributionsCollection"];

    let mut activity: HashMap<String, RepoActivity> = HashMap::new();

    for item in collection["commitContributionsByRepository"]
        .as_array()
        .unwrap_or(&vec![])
    {
        let slug = item["repository"]["nameWithOwner"]
            .as_str()
            .unwrap_or("")
            .to_string();
        if slug.is_empty() {
            continue;
        }
        let private = item["repository"]["isPrivate"].as_bool().unwrap_or(true);
        let count = item["contributions"]["totalCount"].as_u64().unwrap_or(0);
        let entry = activity.entry(slug).or_default();
        entry.commits += count;
        entry.is_private = private;
    }

    for item in collection["pullRequestContributionsByRepository"]
        .as_array()
        .unwrap_or(&vec![])
    {
        let slug = item["repository"]["nameWithOwner"]
            .as_str()
            .unwrap_or("")
            .to_string();
        if slug.is_empty() {
            continue;
        }
        let private = item["repository"]["isPrivate"].as_bool().unwrap_or(true);
        let count = item["contributions"]["totalCount"].as_u64().unwrap_or(0);
        let entry = activity.entry(slug).or_default();
        entry.prs += count;
        entry.is_private = private;
    }

    Ok(activity)
}

/// Fetches the star count for each slug in a single batched GraphQL query.
///
/// Uses aliased `repository` fields (`r0`, `r1`, …) so only one round-trip is
/// needed regardless of list length.  Repos that fail to resolve are silently
/// omitted (the caller falls back to 0 stars).
fn fetch_stars(slugs: &[String]) -> Result<HashMap<String, u64>, Box<dyn std::error::Error>> {
    if slugs.is_empty() {
        return Ok(HashMap::new());
    }

    let fields: Vec<String> = slugs
        .iter()
        .enumerate()
        .filter_map(|(i, slug)| {
            let mut parts = slug.splitn(2, '/');
            let owner = parts.next()?;
            let name = parts.next()?;
            Some(format!(
                "  r{i}: repository(owner: \"{owner}\", name: \"{name}\") {{ stargazerCount }}"
            ))
        })
        .collect();
    let query = format!("{{\n{}\n}}", fields.join("\n"));

    let output = Command::new("gh")
        .args(["api", "graphql", "-f", &format!("query={query}")])
        .output()
        .map_err(|e| format!("Failed to fetch stars: {e}"))?;

    if !output.status.success() {
        return Err(format!(
            "Star fetch failed:\n{}",
            String::from_utf8_lossy(&output.stderr)
        )
        .into());
    }

    let json: serde_json::Value = serde_json::from_slice(&output.stdout)?;
    let data = &json["data"];

    let mut result = HashMap::new();
    for (i, slug) in slugs.iter().enumerate() {
        if let Some(n) = data[format!("r{i}")]["stargazerCount"].as_u64() {
            result.insert(slug.clone(), n);
        }
    }

    Ok(result)
}

// MARK: File patching
/// Replaces the `[[oss_contribs]]` array-of-tables in the TOML document with
/// `entries`, preserving all other keys verbatim.
fn patch_oss_contribs(
    content: &str,
    entries: &[(String, String, u64)],
) -> Result<String, Box<dyn std::error::Error>> {
    let mut doc = content
        .parse::<toml_edit::DocumentMut>()
        .map_err(|e| format!("Could not parse {CONFIG_PATH} as TOML: {e}"))?;

    doc.remove("oss_contribs");

    let mut aot = toml_edit::ArrayOfTables::new();
    for (i, (slug, name, stars)) in entries.iter().enumerate() {
        let mut table = toml_edit::Table::new();
        table["slug"] = toml_edit::value(slug.as_str());
        table["name"] = toml_edit::value(name.as_str());
        table["stars"] = toml_edit::value(*stars as i64);
        let prefix = if i == 0 {
            "\n# Misc OSS contributions - updated by: cargo scripts/update-oss-contribs.rs\n"
        } else {
            "\n"
        };
        table.decor_mut().set_prefix(prefix);
        aot.push(table);
    }
    doc["oss_contribs"] = toml_edit::Item::ArrayOfTables(aot);

    Ok(doc.to_string())
}

// ─── Date helpers ────────────────────────────────────────────────────────────

fn year_ago_rfc3339() -> String {
    let dt = chrono::Utc::now() - chrono::Duration::days(365);
    dt.format("%Y-%m-%dT%H:%M:%SZ").to_string()
}

fn now_rfc3339() -> String {
    chrono::Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string()
}

File: scripts/update-oss-contribs.ts
// Regenerates app/data/oss-contribs.ts from live GitHub contribution data.
//
// Run from the repository root:
//
//   mise run oss-contribs:update
//   # or: aubx tsx scripts/update-oss-contribs.ts
//
// Requires a GitHub token: GITHUB_TOKEN / GH_TOKEN, or an authenticated `gh`
// CLI (the script falls back to `gh auth token`).
//
// It:
//   1. Fetches your commit + PR contributions for the past year (GraphQL).
//   2. Drops own/work-org repos, repos already in PROJECTS, private repos,
//      and repos with no PRs.
//   3. Fetches star counts and sorts by popularity (then slug).
//   4. Preserves any custom display names from the current file.
//   5. Rewrites app/data/oss-contribs.ts.

import { execFileSync } from "node:child_process";
import { writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";
import { OSS_CONTRIBS } from "../app/data/oss-contribs.ts";
import { PROJECTS } from "../app/data/projects.ts";

// MARK: Configuration

const GITHUB_USER = "JP-Ellis";

// Orgs/owners excluded entirely (own repos and work orgs). Repos already in
// PROJECTS are excluded separately regardless of org.
const SKIP_ORGS = new Set(["JP-Ellis", "hep-rs", "pactflow"]);

// Minimum PRs to qualify for the list.
const MIN_PRS = 1;

const DAYS_PER_YEAR = 365;
const MS_PER_DAY = 86_400_000;
const THOUSANDS = 10_000;

const OUT_PATH = join(
  dirname(fileURLToPath(import.meta.url)),
  "..",
  "app/data/oss-contribs.ts",
);

interface RepoActivity {
  commits: number;
  prs: number;
  isPrivate: boolean;
}

// MARK: Token

function resolveToken(): string {
  const fromEnv = process.env.GITHUB_TOKEN ?? process.env.GH_TOKEN;
  if (fromEnv) {
    return fromEnv;
  }
  try {
    return execFileSync("gh", ["auth", "token"], { encoding: "utf8" }).trim();
  } catch (error) {
    throw new Error(
      "No GITHUB_TOKEN/GH_TOKEN set and `gh auth token` failed. " +
        "Export a token or run `gh auth login`.",
      { cause: error },
    );
  }
}

// MARK: GitHub API

async function graphql(
  query: string,
  token: string,
): Promise<Record<string, unknown>> {
  const resp = await fetch("https://api.github.com/graphql", {
    method: "POST",
    headers: {
      // biome-ignore lint/style/useNamingConvention: standard HTTP header name
      Authorization: `Bearer ${token}`,
      "User-Agent": "jpellis-me/1.0",
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ query }),
  });
  if (!resp.ok) {
    throw new Error(`GraphQL status ${resp.status}`);
  }
  const body = (await resp.json()) as Record<string, unknown>;
  return (body.data ?? {}) as Record<string, unknown>;
}

function accumulate(
  activity: Map<string, RepoActivity>,
  list: unknown,
  field: "commits" | "prs",
): void {
  if (!Array.isArray(list)) {
    return;
  }
  for (const raw of list) {
    const item = raw as Record<string, unknown>;
    const repo = item.repository as Record<string, unknown> | undefined;
    const slug = repo?.nameWithOwner;
    if (typeof slug !== "string" || slug === "") {
      // biome-ignore lint/style/noContinue: guard-clause skip for malformed items
      continue;
    }
    const isPrivate = repo?.isPrivate === true;
    const contributions = item.contributions as
      | Record<string, unknown>
      | undefined;
    const count =
      typeof contributions?.totalCount === "number"
        ? contributions.totalCount
        : 0;
    const entry = activity.get(slug) ?? { commits: 0, prs: 0, isPrivate };
    entry[field] += count;
    entry.isPrivate = isPrivate;
    activity.set(slug, entry);
  }
}

async function fetchContributions(
  token: string,
  from: string,
  to: string,
): Promise<Map<string, RepoActivity>> {
  const query = `{
  user(login: "${GITHUB_USER}") {
    contributionsCollection(from: "${from}", to: "${to}") {
      commitContributionsByRepository(maxRepositories: 100) {
        repository { nameWithOwner isPrivate }
        contributions { totalCount }
      }
      pullRequestContributionsByRepository(maxRepositories: 100) {
        repository { nameWithOwner isPrivate }
        contributions { totalCount }
      }
    }
  }
}`;

  const data = await graphql(query, token);
  const user = data.user as Record<string, unknown> | undefined;
  const collection = user?.contributionsCollection as
    | Record<string, unknown>
    | undefined;

  const activity = new Map<string, RepoActivity>();
  accumulate(activity, collection?.commitContributionsByRepository, "commits");
  accumulate(activity, collection?.pullRequestContributionsByRepository, "prs");
  return activity;
}

// Fetches star counts for all slugs in a single aliased GraphQL query.
async function fetchStars(
  token: string,
  slugs: string[],
): Promise<Map<string, number>> {
  if (slugs.length === 0) {
    return new Map();
  }

  const fields = slugs
    .map((slug, i) => {
      const [owner, name] = slug.split("/");
      return owner && name
        ? `  r${i}: repository(owner: "${owner}", name: "${name}") { stargazerCount }`
        : "";
    })
    .filter((field) => field !== "")
    .join("\n");
  const data = await graphql(`{\n${fields}\n}`, token);

  const stars = new Map<string, number>();
  slugs.forEach((slug, i) => {
    const node = data[`r${i}`] as Record<string, unknown> | undefined;
    const count = node?.stargazerCount;
    if (typeof count === "number") {
      stars.set(slug, count);
    }
  });
  return stars;
}

// MARK: Output

function numberLiteral(n: number): string {
  // Match the repo style: group thousands for 5+ digit numbers.
  return n < THOUSANDS
    ? String(n)
    : n.toLocaleString("en-US").replace(/,/gu, "_");
}

function renderFile(
  entries: { slug: string; name: string; stars: number }[],
): string {
  const lines = entries.map(
    (e) =>
      `  { slug: ${JSON.stringify(e.slug)}, name: ${JSON.stringify(e.name)}, stars: ${numberLiteral(e.stars)} },`,
  );
  return `export const OSS_CONTRIBS: { slug: string; name: string; stars: number }[] = [\n${lines.join("\n")}\n];\n`;
}

// MARK: Main

async function main(): Promise<void> {
  const token = resolveToken();

  const tracked = new Set(
    PROJECTS.flatMap((project) =>
      project.link?.kind === "github" ? [project.link.slug] : [],
    ),
  );
  const existingNames = new Map(OSS_CONTRIBS.map((c) => [c.slug, c.name]));

  const now = Date.now();
  const from = new Date(now - DAYS_PER_YEAR * MS_PER_DAY).toISOString();
  const to = new Date(now).toISOString();

  console.error(`Fetching contributions ${from} → ${to} for @${GITHUB_USER}…`);
  const activity = await fetchContributions(token, from, to);

  const candidates = [...activity.entries()]
    .filter(([slug, data]) => {
      const org = slug.split("/")[0] ?? "";
      return (
        !(SKIP_ORGS.has(org) || tracked.has(slug) || data.isPrivate) &&
        data.prs >= MIN_PRS
      );
    })
    .map(([slug]) => slug);

  console.error(`Fetching star counts for ${candidates.length} repos…`);
  const stars = await fetchStars(token, candidates);

  candidates.sort((a, b) => {
    const byStars = (stars.get(b) ?? 0) - (stars.get(a) ?? 0);
    return byStars === 0 ? a.localeCompare(b) : byStars;
  });

  const entries = candidates.map((slug) => ({
    slug,
    name: existingNames.get(slug) ?? slug.split("/")[1] ?? slug,
    stars: stars.get(slug) ?? 0,
  }));

  await writeFile(OUT_PATH, renderFile(entries));

  console.error(
    `Updated app/data/oss-contribs.ts — ${entries.length} entries:`,
  );
  for (const entry of entries) {
    const isNew = existingNames.has(entry.slug) ? "" : "  [new]";
    console.error(`  ${entry.slug} (${entry.stars}★)${isNew}`);
  }
}

main().catch((error: unknown) => {
  console.error(error);
  process.exit(1);
});
